File: classic-algorithm-in-action/classic-algorithm-solution-in-java/src/com/IntegerArray/SubarraySumZero.java
package com.IntegerArray;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class SubarraySumZero {
	/**
     * @param nums: A list of integers
     * @return: A list of integers includes the index of the first number 
     *          and the index of the last number
     */
    public ArrayList<Integer> subarraySum(int[] nums) {
        ArrayList<Integer> res = new ArrayList<>();
		for (int i = 0; i < nums.length; i++) {
			int thisSum = 0;
			for (int j = i; j < nums.length; j++){
				thisSum += nums[j];
				if (thisSum == 0) {
					res.add(i);
					res.add(j);
					return res;
				}
			}
		}
		return res;
    }
    
    public ArrayList<Integer> subarraySumBetter(int[] nums) {
        ArrayList<Integer> res = new ArrayList<>();
		Map<Integer, Integer> map = new HashMap<>(); // key:sum, value:index
		map.put(0, 0);
		
		int curr_sum = 0;
		for (int i = 0; i < nums.length; i++) {
			curr_sum += nums[i];
			if (map.containsKey(curr_sum)) {
				res.add(map.get(curr_sum));
				res.add(i);
				break;
			} else {
				map.put(curr_sum, i+1);
			};
		}
		return res;
    }
    
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		SubarraySumZero s = new SubarraySumZero();
		int[] nums = {-3, 1, 2, -3, 4};
		System.out.println(s.subarraySum(nums));
		System.out.println(s.subarraySumBetter(nums));
	}

}

File: classic-algorithm-in-action/classic-algorithm-solution-in-python/SubarraySumZero.py
# -*- coding: utf-8 -*-
"""
Created on Tue Mar 20 09:44:04 2018

@problem: lintcode 138. Subarray Sum

@author: sheng
"""

class Solution:
    """
    @param nums: A list of integers
    @return: A list of integers includes the index of the first number and the index of the last number
    """
    def subarraySum(self, nums):
        # LTE!
        for i in range(len(nums)):
            for j in range(len(nums)):
                if sum(nums[i:j+1]) == 0:
                    return [i, j]
    
    def subarraySumBetter(self, nums):
        res = []
        d = {}
        d[0] = 0
        
        curr_sum = 0
        for i, num in enumerate(nums):
            curr_sum += num
            if curr_sum in d:
                res.append(d[curr_sum])
                res.append(i)
                break
            else:
                d[curr_sum] = i+1
        return res
    
def main():
    s = Solution()
    nums = [-3, 1, 2, -3, 4]
    print(s.subarraySum(nums))
    print(s.subarraySumBetter(nums))
    
if __name__ == '__main__':
    main()